nomos-mix/core/src/processor/mod.rs

```diff
@@ -0,0 +1,72 @@
+mod crypto;
+mod temporal;
+
+pub use crypto::CryptographicProcessorSettings;
+pub use temporal::TemporalProcessorSettings;
+
+use serde::{Deserialize, Serialize};
+use tokio::sync::mpsc;
+
+use crate::processor::crypto::CryptographicProcessor;
+
+#[derive(Clone, Debug, Serialize, Deserialize, Default)]
+pub struct ProcessorSettings {
+    pub cryptographic_processor: CryptographicProcessorSettings,
+    pub temporal_processor: TemporalProcessorSettings,
+}
+
+/// [`Processor`] implements "Message Router" defined in the Tier-2 spec.
+/// - Wraps new messages using [`CryptographicProcessor`]
+/// - Unwraps incoming messages received from network using [`CryptographicProcessor`]
+/// - Pushes unwrapped messages to [`TemporalProcessor`] to delay them
+/// - Releases messages returned by [`TemporalProcessor`]
+pub struct Processor {
```

Comment:
Not claiming that it is a correct way but have you considered having only `inbound_message_processor` and `outbound_message_processor` and construct a message as `{payload, routing_data}`, where `routing_data` defines the context on how the message should be processed?

Reply:
I think I have misunderstood this part. Can you confirm that you are making the Message Router a "governor" or a primary entity that encapsulates cryptographic and temporal processors? It is a bit different than I have imagined. My vision was that the Message Router will be "called" by the Cryptographic Processor, and the CP is called by Tier 1 or Tier 3. I was assuming that neither CP or MR (or even TP) are going to have any clocks that determine their work, they should just process messages when needed. We are going to have clocks at Tier 1 and Tier 3.

Regarding TP: the implementation of the temporal processor might not physically add any delay but add a release time to the message metadata and pass it to the PTM. The PTM would read messages from the input queue when the release time is met.

Reply:
> I think I have misunderstood this part. Can you confirm that you are making the Message Router a "governor" or a primary entity that encapsulates cryptographic and temporal processors?

Yes. I'll rename the `Processor` to `MessageBlender`, which is the name of Tier 2 defined in the spec. And, I'll explicitly define the `MessageRouter`. The `MessageBlender` will use it. 

>  I was assuming that neither CP or MR (or even TP) are going to have any clocks that determine their work, they should just process messages when needed. We are going to have clocks at Tier 1 and Tier 3.

Yes. CP doesn't have clocks in this implementation. Also, the part that does the same things as MR also doesn't have clocks. But, as I mentioned above, I will refactor this module to make it look more identical to the spec.

> Regarding TP: the implementation of the temporal processor might not physically add any delay but add a release time to the message metadata and pass it to the PTM. The PTM would read messages from the input queue when the release time is met.

Yes. The code comment here is just for explaining the role of TP shortly. If you think this comment is not correct, I will remove it (the `to delay them` part).

Reply:
I renamed `Processor` to `MessageBlend`. It's much clearer. 

But still, I'm not defining the `MessageRouter` module in this code. The role of `MessageRouter` is very very simple. The `MessageBlend` can perform the role without any overhead. The code looks simple and easy. Even lighter in term of computations.
If we define the `MessageRouter` explicitly in the code, we'll end up adding more async channels and data structures to make proper orders to `MessageRouter`. I think it will make code verbose.
As long as this code implements what `MessageRouter` should do correctly, I don't think the implementation must follow the structure defined in the spec 100% the same.